# Optimization 4 - Constrained Optimization

In [1]:
import numpy as np

rng = np.random.default_rng(seed=42)

## 1 The Constraint

### 1.1 Loss Function

We use Gradient Descent algorithm to find best weight vector $w^*, w_0^*$ by optimizing/minimizing the Loss function.

#### Optimization of Loss Function

Equation for **Optimization of Loss Function** for Binary Classification can be expanded as:

$$
\large
\begin{aligned}
w^*, w_0^* &= \arg\min_{w, w_0} \;\; l(D, \vec{w}, w_0) \\[10pt]
           &= \arg\min_{w, w_0}\;-\frac{1}{n}\sum_{i=1}^n\Biggl(\frac{w^T x^{(i)}\;+\;w_0}{\Vert{w}\Vert}\Biggr)\cdot y^{(i)} \\[10pt]
           &= \arg\min_{w, w_0}\;-\frac{1}{n}\sum_{i=1}^n\Biggl(\frac{w^T x^{(i)}\;+\;w_0}{\sqrt{w_1^2 + w_2^2 + \ldots + w_n^2}}\Biggr)\cdot y^{(i)} \\[10pt]
           &= \arg\min_{w, w_0}\;-\frac{1}{n}\sum_{i=1}^n\Biggl(\frac{w^T x^{(i)}\;+\;w_0}{\sqrt{\vec{w}^2}}\Biggr)\cdot y^{(i)}
\end{aligned}
$$

### 1.2 Derivative of Loss Function

Given loss function:

$$
\large
\begin{aligned}
l(D, \vec{w}, w_0) = -\frac{1}{n}\sum_{i=1}^n\Biggl(\frac{w^T x^{(i)}\;+\;w_0}{\Vert{w}\Vert}\Biggr)\cdot y^{(i)}
\end{aligned}
$$

Discarding $y$, function to represent optimization of a single data-point can be written as:

$$
\large
\begin{aligned}
f(\vec{w}, w_0) = \frac{w_1 x_1 + w_2 x_2 + \ldots +  w_d x_d + w_0}{\sqrt{w_1^2 + w_2^2 + \ldots + w_d^2}}
\end{aligned}
$$

The partial derivatives of $f(\vec{w}, w_0)$ can be calculated as:

$$
\large
\begin{aligned}
\frac{\partial}{\partial{\vec{w}}} f(\vec{w}, w_0) \\[8pt]
\frac{\partial}{\partial{w_0}} f(\vec{w}, w_0)
\end{aligned}
$$

Above equation represents the form:

$$
\large
\begin{aligned}
h(w) = \frac{f(w)}{g(w)} = \frac{w_1 x_1 + w_2 x_2 + \ldots +  w_d x_d + w_0}{\sqrt{w_1^2 + w_2^2 + \ldots + w_d^2}}
\end{aligned}
$$

To get derivatives of above function $h(x)$ we can apply Quotient rule given below:

$$
\large
\begin{aligned}
\frac{d}{dx} \Biggl[\frac{f(x)}{g(x)}\Biggr] = \frac{g(x)\frac{d}{dx}f(x) - f(x)\frac{d}{dx}g(x)}{\bigl[g(x)\bigr]^2}
\end{aligned}
$$

But we know that:

$$
\large
\begin{aligned}
\frac{\partial}{\partial\vec{w}} (w_1 x_1 + \ldots +  w_d x_d + w_0) = \vec{x}
\end{aligned}
$$

$$
\large
\begin{aligned}
\nabla_{\vec{w}}\;(w^T{x} + w_0) = \begin{bmatrix} x_1 \\ x_2 \\ \vdots \\ x_d \end{bmatrix}
\end{aligned}
$$

i.e., Gradient Vector of weight vector $\vec{w}$ is its feature vector $\vec{x}$.

and also know that:

$$
\large
\begin{aligned}
\frac{\partial}{\partial\vec{x}} (x^T x) = 2\cdot\vec{x}
\end{aligned}
$$

$$
\large
\begin{aligned}
\nabla_{\vec{x}}\;(x^T x) = 2 \times \begin{bmatrix} x_1 \\ x_2 \\ \vdots \\ x_d \end{bmatrix}
\end{aligned}
$$

i.e., Gradient Vector of a dot product between same vectors is a scalar product between 2 and that vector itself.

### 1.3 Complex Minimization

#### The Constraint

Finding derivatives for below equation is very complex:

$$
\large
\begin{aligned}
\frac{w_1 x_1 + w_2 x_2 + \ldots +  w_d x_d + w_0}{\sqrt{w_1^2 + w_2^2 + \ldots + w_d^2}}
\end{aligned}
$$

Here the problem is denominator, which can be simplified by making some assumptions (or setting some constraints).

Based on our understanding below:

- We know that **Gradient vector** $\nabla f$ at any given point represents the **steepness** (towards ascent) and **direction** (towards maxima).
- Therefore **negative** of $\nabla f$ vector i.e., $-\nabla f$ Vector automatically points in the **direction of minima**.
- It is only useful to know the direction of $-\nabla f$ vector and its **magnitude has no significance**.

We can set a below constrain to simplify the equation:

$$
\large
\begin{aligned}
\boxed{
\rule{0pt}{15pt} \\
\quad \Vert\vec{w}\Vert = 1 \quad \\[10pt]
}
\end{aligned}
$$

i.e., the magnitude of weight vector will always be equal to 1.

### 1.4 Example

#### Example #1

Find the minima of given $f(x)$, such that it satisfies the constraint: $2x^2 - 3x + 1 = 0$

$$
\large
\begin{aligned}
f(x) &= x^2 - 3x + 3
\end{aligned}
$$

##### PART 1: Minimization

**STEP 1**: Compute first derivative.

$$
\large
\begin{aligned}
f^\prime(x) = 2x - 3
\end{aligned}
$$

**STEP 2**: Find Critical value.

$$
\large
\begin{aligned}
f^\prime(x) &= 0 \\[5pt]
     2x - 3 &= 0 \\[5pt]
          x &= \frac{3}{2} \\[5pt]
            &= 1.5  
\end{aligned}
$$

**STEP 3**: Identify Minima

$$
\large
\begin{aligned}
f^{\prime\prime}(x) &= 2 \\[5pt]
                  2 &> 0
\end{aligned}
$$

Hence $f(x)$ takes minimum value at critical value of $x$:

$$
\large
\begin{aligned}
x = 1.5
\end{aligned}
$$

**STEP 4**: Calculate minima value.

$$
\large
\begin{aligned}
f(x) &= \biggl(\frac{3}{2}\biggr)^2 - 3 \biggl(\frac{3}{2}\biggr) + 3 \\[10pt]
     &= 0.75
\end{aligned}
$$

Hence minima is located at: $\bigl(x, f(x)\bigr) = (1.5, 0.75)$

##### PART 2: Constraint Optimization

Plotting the constraint: $2x^2 - 3x + 1$ and $f(x) = x^2 - 3x + 3$ and solving using visualization:

https://www.desmos.com/calculator/aaurlqyzhj

Both the equation intersect at At $\bigl(x, f(x)\bigr) = (1.414, 0.757) \approx (1.5, 0.75)$

### 1.5 Summary

- In order to minimize the complexity in calculating the partial derivatives of loss function certain constrains are introduced.
- Since there is no significance of the magnitude of a Gradient vector, a constraint is set to always equate it to 1.
- Setting a constraint alters the Optimization process followed earlier and requires **Constrain Optimization**.

## 2 Constraint Optimization

### 2.1 New Optimization Problem

#### What is Constraint Optimization?

Constraint Optimization is the process of loss minimization of a function such that the target constraint is satisfied.

New Optimization equation has two parts loss minimization and target constraint:

$$
\large
\begin{aligned}
w^*, w_0^* &= \arg\min_{\vec{w},w_0}\;\;l(D, \vec{w},w_0) & \text{such that}:\;\Vert{\vec{w}}\Vert = 1 \\[10pt]
           &= \arg\min_{\vec{w},w_0}\;-\frac{1}{n}\sum_{i=1}^n\Biggl(\frac{w^T x^{(i)}\;+\;w_0}{\Vert{w}\Vert}\Biggr)y^{(i)} & \text{such that}:\;\Vert{\vec{w}}\Vert = 1
\end{aligned}
$$

Above equation can be solved using Lagrange Function.

### 2.2 Lagrangian Function

#### Lagrange Multiplier

$$
\large
\begin{aligned}
\mathcal{L}(x, \lambda) \equiv f(x)\;+\;\lambda\cdot{g(x)}
\end{aligned}
$$

Where $\lambda$ is called as Lagrange Multiplier.

#### Lagrange Multiplier for Constraint Optimization

Applying Lagrange Multiplier on Optimization equation:

$$
\large
\begin{aligned}
w^*, w_0^* &= \arg\min_{\vec{w},w_0,\lambda}\;\;l\bigl(D,\vec{w},w_0,\lambda\bigr) \\[10pt]
           &= \arg\min_{\vec{w},w_0,\lambda}\;-\frac{1}{n}\sum_{i=1}^n\Bigl(w^T x^{(i)}\;+\;w_0\Bigr)y^{(i)}\;+\;\lambda \bigl(\Vert{\vec{w}}\Vert - 1\bigr)
\end{aligned}
$$

##### Why magnitude of w vector minus 1?

$$
\large
\begin{aligned}
    \Vert{\vec{w}}\Vert &= 1 \\[5pt]
\Vert{\vec{w}}\Vert - 1 &= 0
\end{aligned}
$$

### 2.3 Examples

#### Example 1

### 2.4 GD Algorithm with LM

#### Old GD Update Rule

$$
\large
\begin{aligned}
w^n = x^o - \eta \frac{\partial}{\partial\;{w}} f(w, w_0) \\[8pt]
w_0^n = w_0^o - \eta \frac{\partial}{\partial\;{w_0}} f(w, w_0)
\end{aligned}
$$

Where,

- $w$ - Weight vector
- $w_0$ - Bias

#### New GD Update Rule with LM

$$
\large
\begin{aligned}
w^n = x^o - \eta \frac{\partial}{\partial\;{w}} f(w, w_0, \lambda) \\[8pt]
w_0^n = w_0^o - \eta \frac{\partial}{\partial\;{w_0}} f(w, w_0, \lambda)
\end{aligned}
$$

Where,

- $w$ - Weight vector
- $w_0$ - Bias

$$
\large
\begin{aligned}
\vec{w}^{t+1} &= \vec{w}^t - \eta\Biggl[-\sum_{i=1}^n x_i \cdot y_i + \lambda\frac{\vec{w}^t}{\Vert{\vec{w}^t}\Vert}\Biggr] \\[10pt]
    w_0^{t+1} &= w_0^t - \eta\Biggl[-\sum_{i=1}^n y_i\Biggr]
\end{aligned}
$$

Where,

$\lambda$ is the regularizer

> **Note**:
>
> 1. Gradient of weight vector $w$ before Constraint Optimization was feature vector $x$
> 2. Gradient of weight vector $w$ after Constraint Optimization was feature vector $x \cdot y$

## 3 Flavors of Gradient Descent

### 3.1 Epoch

#### What is an Epoch?

One epoch is defined as one complete iteration of entire dataset.

Given a dataset $D$ where $D$ is defined as:

$$
\large
\begin{aligned}
D := \biggl\{\bigl(x_i, y_i \bigr)\;;\;x\in\mathbb{R}^d\;;\;y\in\{-1,+1\}\biggr\}_{i=1}^n
\end{aligned}
$$

One epoch is when all $n$ rows in the dataset $D$ are used for some calculation.

### 3.2 Types of Gradient Descent

There are three types of Gradient Descent Algorithm based on number of rows used in weights calculation:

1. Vanilla Gradient Descent
2. Stochastic Gradient Descent
3. Batch Gradient Descent

#### 1 Vanilla Gradient Descent

##### Properties

$$
\large
\begin{aligned}
\vec{w}^{t+1} = \vec{w}^t - \eta\Biggl[-\sum_{i=1}^n x_i \cdot y_i + \lambda\frac{\vec{w}^t}{\Vert{\vec{w}^t}\Vert}\Biggr]
\end{aligned}
$$

- Entire dataset is used to compute the weights.
- Weights are updated after computing gradients using **all rows** from the dataset.
- Weights are updated **only once per Epoch**.

##### Advantages

Most **accurate** of all the three variants of GD since entire dataset is used to updated weights.

##### Disadvantages

Training process is **inefficient** consumes lot of time and compute power since entire dataset is used each time while computing new weights.

#### 2 Stochastic Gradient Descent

##### Properties

$$
\large
\begin{aligned}
\vec{w}^{t+1} = \vec{w}^t - \eta\Biggl[- (x_i \cdot y_i) + \lambda\frac{\vec{w}^t}{\Vert{\vec{w}^t}\Vert}\Biggr]
\end{aligned}
$$

- One row is used at a time to compute the weights.
- Weights are updated immediately after computing gradients using **single row** from the dataset.
- Weights are updated **$n$ times in an Epoch**.

##### Advantages

Most **efficient** of all the three variants of GD since only one dataset is used to updated weights.

##### Disadvantages

Results in **zigzag** Learning i.e., one odd datapoint can diverge/subvert the loss minimization towards minima.

#### 3 Mini-Batch Gradient Descent

##### Properties

$$
\large
\begin{aligned}
\vec{w}^{t+1} = \vec{w}^t - \eta\Biggl[-\sum_{i=1}^b x_i \cdot y_i + \lambda\frac{\vec{w}^t}{\Vert{\vec{w}^t}\Vert}\Biggr]
\end{aligned}
$$

Where $b < n$

- Dataset is divided into $k$ number of batches where each batch is of size $b$ where $b < n$.
- $b$ number of rows are used to compute the weights.
- Weights are updated $k$ number of times per Epoch.
- Adds new Hyper-parameter batch size $b$ for tuning.

##### Advantages

- Mini-Batch Gradient Descent is more efficient when compared with Vanilla Gradient Descent.
- Mini-Batch Gradient Descent is more accurate when compared with Stochastic Gradient Descent.

##### Disadvantages

- Mini-Batch Gradient Descent is less accurate when compared with Vanilla Gradient Descent.
- Mini-Batch Gradient Descent is less efficient when compared with Stochastic Gradient Descent.

### 3.3 Examples

#### Test Dataset

In [2]:
f1 = [41, 20, 30, 47, 40, 50, 25, 28, -2000, 5000]
cl = [+1, -1, -1, +1, +1, +1, -1, -1, +1, -1]

n = len(f1)
X = np.array(f1).reshape((n, 1))
Y = np.array(cl).reshape((n, 1))
data = np.concat((X, Y), axis=1)
rng.shuffle(data)
data, data.shape

(array([[   50,     1],
        [   25,    -1],
        [   41,     1],
        [   28,    -1],
        [   47,     1],
        [   30,    -1],
        [   40,     1],
        [ 5000,    -1],
        [   20,    -1],
        [-2000,     1]]),
 (10, 2))

In [3]:
# Hyper-parameters.
eta = 0.001
lamda = 1000

In [4]:
def gradient_descent(one_epoch):
    """
    Simple implementation of GD.
    """
    convergence = False
    w_old = 0
    w_new = None
    itr = 0
    while not convergence:
        # Apply Gradient Update Rule.
        w_new = one_epoch(w_old)

        # Check for convergence.
        if np.allclose(w_new, w_old, rtol=1e-08, atol=1e-08):
            convergence = True

        w_old = w_new
        itr += 1

    return w_new, itr

#### Example #1

##### Vanilla Gradient Descent

In [5]:
def one_epoch_vgd(w_t):
    """
    One epoch of Vanilla Gradient Descent.
    """
    w_tp1 = w_t - eta * (-np.sum(X * Y) + (lamda * w_t))
    return w_tp1


w, itr = gradient_descent(one_epoch_vgd)
print(f"New weight after {itr} epochs: [{w.round(6).item()}]")

New weight after 2 epochs: [-6.925]


##### Stochastic Gradient Descent

In [6]:
def one_epoch_sgd(w_t):
    """
    One epoch of Stochastic Gradient Descent.
    """
    for grad in X * Y:
        w_tp1 = w_t - eta * (-grad + (lamda * w_t))
        w_t = w_tp1

    return w_tp1


w, itr = gradient_descent(one_epoch_sgd)
print(f"New weight after {itr} epochs: [{w.round(6).item()}]")

New weight after 2 epochs: [-5.0]


##### Batch Gradient Descent

In [7]:
def one_epoch_bgd(w_t):
    """
    One epoch of Mini-Batch Gradient Descent.
    """
    b = 5  # Batch size
    k = np.ceil(n / b).item()

    # One epoch
    for batch in np.array_split((X * Y), k):
        w_tp1 = w_t - eta * (-np.sum(batch) + (lamda * w_t))
        w_t = w_tp1

    return w_tp1


w, itr = gradient_descent(one_epoch_bgd)
print(f"New weight after {itr} epochs: [{w.round(6).item()}]")

New weight after 2 epochs: [-7.003]


#### Observations

- Data-points: `[[ 5000, -1], [-2000, 1]]` are outliers in dataset.
- Weights computed using Mini-Batch Gradient Descent: `[-7.0]` is closer to weights computed using Vanilla Gradient Descent: `[-6.925]`.
- Stochastic Gradient Descent was affected by outliers in datatset.

## 4 Implementation

Complete implementation of Mini-Batch Gradient Descent in Python.